Add tests on the Dataset output

satyaog opened the pull request:
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes #[ISSUE_NUMBER]

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

# Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

satyaog committed:
Add tests on the Dataset output

satyaog commented:
Some of the tests are failing most likely because of a misundertanding on my part but if you could quickly clear them (marked as TODOs in the comments), I would work to fix the code. Do that when you can, there are still some obvious ones that I can work on

mweiss17 commented on tests/models_test.py:
```diff
@@ -44,22 +45,28 @@ def test_run(self):
+            # TODO: Fix this test which is failing because preprocessing doesn't produce
+            #  the right number of days apparently
+            # self.assertEqual(len(days_output), n_days)
+
```
if there are no encounters, then we might not have any logs for that day. A solution would be to instead of looking at the encounter logs to find the start and end date of the simulation, to look at the "daily_logs"

satyaog replied:
True, last commit contains a small fix for the time being which should work most of the time.

mweiss17 commented on tests/models_test.py:
```diff
@@ -44,22 +45,28 @@ def test_run(self):
                     with open(pkl, 'rb') as f:
                         day_humans.append(pickle.load(f))
-                output.append(day_humans)
+                # TODO: Fix this test which is failing because simulation or preprocessing
```
This sounds like a bug, too. We should have the risk pred code running every day for every person.  I'm not sure why this is happening but we can look into it.

mweiss17 replied:
that said, it might not break downstream tasks this moment, if they are treating each human/day data as a seperate example

satyaog replied:
Apparently they take the ceil of a ratio of the total number of humans for each categories so it's apparently a feature. With a sufficiently high number of humans the gap should go away but I'm testing with only 35 ppl at the moment.

mweiss17 commented on tests/models_test.py:
```diff
@@ -87,24 +94,39 @@ def test_run(self):
+                        self.assertLess(observed['candidate_encounters'][:, 0].max(), 256)
                         self.assertLess(observed['candidate_encounters'][:, 1].max(), 16)
+                        self.assertGreaterEqual(observed['candidate_encounters'][:, 1].min(), 0)
+                        # TODO: Tix this test which is failing because observed['candidate_encounters'][:, 2] as apparently
```
Yea, that should never happen

satyaog replied:
Last commit contains a quick attempt to fix this. I'm not sure to fully grasp the logic of the code yet tho so it might be wrong.

mweiss17 commented on tests/models_test.py:
```diff
@@ -146,34 +168,71 @@ def test_run(self):
 
-                    if unobserved['is_infectious'] or unobserved['is_recovered']:
+                    # observed['reported_symptoms'] is a subset of unobserved['true_symptoms']
+                    # TODO: Fix this test which is failing with reported_symptoms having more symptoms than true_symptoms.
```
this should never happen

mweiss17 commented on tests/models_test.py:
```diff
@@ -146,34 +168,71 @@ def test_run(self):
+                    # self.assertTrue((unobserved['true_symptoms'] == observed['reported_symptoms'])
+                    #                 [observed['reported_symptoms'].astype(np.bool)].all())
+
+                    # TODO: Both unobserved['is_infectious'] and unobserved['is_recovered'] can apparently be True. Is this a bug
```
this also should never happen -- could be an off by one?

mweiss17 commented on tests/models_test.py:
```diff
@@ -146,34 +168,71 @@ def test_run(self):
                         self.assertTrue(unobserved['is_infectious'] != unobserved['is_recovered'])
 
                     # exposed_locs is the same length as candidate_locs
                     # TODO: observed['candidate_locs'] should be a tuple (human_readable, id) preferably sorted
```
we should simply remove the candidate_locs functionality. It is not needed downstream in the foreseeable future, and will never be observable in the real data we receive.

mweiss17 commented on tests/models_test.py:
```diff
@@ -146,34 +168,71 @@ def test_run(self):
+                        # self.assertTrue((unobserved['true_symptoms'][1:] == prev_unobserved['true_symptoms'][:13]).all())
                         self.assertTrue(unobserved['is_exposed'] if prev_unobserved['is_exposed'] else True)
-                        self.assertTrue((unobserved['infectiousness'][:13, :] == prev_unobserved['infectiousness'][-13:, :]).all())
+                        # TODO: Fix this test which is failing because previous and current exactly the same
```
this is expected behaviour. We are modelling infectiousness from viral load, and we are modelling viral load as a piecewise linear function with a rise, plateau, and fall. There are several parameters which determine the shape of this function, but it should suffice to say that this is in line with the current model the epidemiologists have approved

satyaog replied:
But the infectiousness should shift by 1 day right? Here the test fails because the infectiousness array of the last 14 days for say day 14 (with peak at day 10) is the exact same infectiousness array of the last 14 days for day 15 (with peak this time at day 11).

mweiss17 replied:
oh snap
that's not right

satyaog committed:
Fix models tests

Fixed number of days outputed
Attempt to fix the old risk being bigger than a 4 bits int
